jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -5,80 +5,319 @@
 package akka.stream.alpakka.jms
 
 import java.util.concurrent.Semaphore
+import java.util.concurrent.atomic.AtomicBoolean
 import javax.jms._
 
-import akka.stream.stage.{GraphStage, GraphStageLogic, OutHandler, StageLogging}
-import akka.stream.{ActorAttributes, Attributes, Outlet, SourceShape}
+import akka.stream.stage._
+import akka.stream._
 
+import scala.annotation.tailrec
 import scala.collection.mutable
+import scala.concurrent.Future
+import scala.util.control.NonFatal
 import scala.util.{Failure, Success}
 
-final class JmsSourceStage(settings: JmsSourceSettings) extends GraphStage[SourceShape[Message]] {
+final class JmsSourceStage(settings: JmsSourceSettings)
+    extends GraphStageWithMaterializedValue[SourceShape[Message], KillSwitch] {
 
   private val out = Outlet[Message]("JmsSource.out")
 
   override def shape: SourceShape[Message] = SourceShape[Message](out)
 
-  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
-    new GraphStageLogic(shape) with JmsConnector with StageLogging {
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, KillSwitch) = {
+    val logic = new NoAckLogic(shape, out, settings, inheritedAttributes)
```

Comment:
I think putting the graph stage logic inline here and accessing the killswitch with reflective access instead of having an explicit class is fine. Following the logic around here gets a bit convoluted for no obvious reason like this.

Reply:
Let me move these in. You're right, it is better to have it inline.

Reply:
This comment has not yet been addressed for this round, for sanity of the reviewers 😄 since it will be a restructure/refactor that hides the other previous comments. I've addressed pretty much every other comment.